### PR TITLE
Supporting KeyVault Secrets in Application Gateway

### PR DESCRIPTION
**Pull request by akshaysngupta**
## Description
Cmdlet Review PR: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/213

## Checklist

- [ ] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#updating-the-change-log)
- [ ] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (Microsoft internal partners only)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/help-generation.md)

### CODE REVIEW

File: src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj [hunk 1]
```diff
@@ -65,6 +68,15 @@
     <Reference Include="Microsoft.Azure.Management.ContainerInstance">
       <HintPath>..\..\..\packages\Microsoft.Azure.Management.ContainerInstance.1.1.0\lib\net452\Microsoft.Azure.Management.ContainerInstance.dll</HintPath>
     </Reference>
+    <Reference Include="Microsoft.Azure.Test.HttpRecorder, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj]: @akshaysngupta you can remove this chunk of references that was added since they are brought in from a common file elsewhere

File: src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj [hunk 2]
```diff
@@ -43,8 +43,11 @@
     <RunCodeAnalysis>false</RunCodeAnalysis>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="Microsoft.Azure.Management.Network, Version=19.5.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.19.7.0-preview\lib\net452\Microsoft.Azure.Management.Network.dll</HintPath>
+    <Reference Include="Microsoft.Azure.Management.KeyVault, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.KeyVault.2.4.3\lib\net452\Microsoft.Azure.Management.KeyVault.dll</HintPath>
+    </Reference>
+    <Reference Include="Microsoft.Azure.Management.Network, Version=19.6.2.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.19.6.2-preview\lib\net452\Microsoft.Azure.Management.Network.dll</HintPath>
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj]: @akshaysngupta the version of the Network SDK across all projects is 19.7.0-preview, is there any reason to downgrade it here?

File: src/ResourceManager/Network/Commands.Network.Test/packages.config [hunk 3]
```diff
@@ -9,19 +9,22 @@
   <package id="Microsoft.Azure.Management.Authorization" version="2.0.0" targetFramework="net452" />
   <package id="Microsoft.Azure.Management.Compute" version="23.2.0" targetFramework="net452" />
   <package id="Microsoft.Azure.Management.ContainerInstance" version="1.1.0" targetFramework="net452" />
-  <package id="Microsoft.Azure.Management.Network" version="19.7.0-preview" targetFramework="net452" />
+  <package id="Microsoft.Azure.Management.KeyVault" version="2.4.3" targetFramework="net452" />
+  <package id="Microsoft.Azure.Management.Network" version="19.6.2-preview" targetFramework="net452" />
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network.Test/packages.config]: @akshaysngupta same comment here

File: src/ResourceManager/Network/Commands.Network.Test/packages.config [hunk 4]
```diff
@@ -9,19 +9,22 @@
...
   <package id="Microsoft.Azure.Management.Storage" version="6.1.0-preview" targetFramework="net45" />
   <package id="Microsoft.Azure.Test.Framework" version="3.0.0" targetFramework="net452" />
-  <package id="Microsoft.Azure.Test.HttpRecorder" version="1.6.8" targetFramework="net45" />
+  <package id="Microsoft.Azure.Test.HttpRecorder" version="1.13.0" targetFramework="net452" />
   <package id="Microsoft.Bcl" version="1.1.9" targetFramework="net45" />
   <package id="Microsoft.Bcl.Async" version="1.0.168" targetFramework="net45" />
   <package id="Microsoft.Bcl.Build" version="1.0.14" targetFramework="net45" />
   <package id="Microsoft.IdentityModel.Clients.ActiveDirectory" version="2.28.3" targetFramework="net45" />
   <package id="Microsoft.Net.Http" version="2.2.28" targetFramework="net45" />
+  <package id="Microsoft.Rest.ClientRuntime" version="2.3.18" targetFramework="net452" />
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network.Test/packages.config]: @akshaysngupta same comment here

File: src/ResourceManager/Network/Commands.Network.Test/packages.config [hunk 5]
```diff
@@ -9,19 +9,22 @@
...
+  <package id="Microsoft.Azure.Management.KeyVault" version="2.4.3" targetFramework="net452" />
+  <package id="Microsoft.Azure.Management.Network" version="19.6.2-preview" targetFramework="net452" />
   <package id="Microsoft.Azure.Management.OperationalInsights" version="0.19.0-preview" targetFramework="net452" />
   <package id="Microsoft.Azure.Management.Redis" version="3.1.1-preview" targetFramework="net45" />
   <package id="Microsoft.Azure.Management.ResourceManager" version="1.9.1-preview" targetFramework="net452" />
   <package id="Microsoft.Azure.Management.Resources" version="2.20.0-preview" targetFramework="net45" />
   <package id="Microsoft.Azure.Management.Storage" version="6.1.0-preview" targetFramework="net45" />
   <package id="Microsoft.Azure.Test.Framework" version="3.0.0" targetFramework="net452" />
-  <package id="Microsoft.Azure.Test.HttpRecorder" version="1.6.8" targetFramework="net45" />
+  <package id="Microsoft.Azure.Test.HttpRecorder" version="1.13.0" targetFramework="net452" />
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network.Test/packages.config]: @akshaysngupta you should be able to revert this change, and even remove this reference to `HttpRecorder` since we pull it from NuGet elsewhere

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs [hunk 6]
```diff
@@ -183,6 +183,13 @@ public class NewAzureApplicationGatewayCommand : ApplicationGatewayBaseCmdlet
             HelpMessage = "A hashtable which represents resource tags.")]
         public Hashtable Tag { get; set; }
 
+        [Parameter(
+            Mandatory = true,
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs]: @akshaysngupta adding a mandatory parameter to an existing cmdlet is a breaking change. You will need to make this optional.

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/SslCertificate/AzureApplicationGatewaySslCertificateBase.cs [hunk 7]
```diff
@@ -29,24 +30,50 @@ public class AzureApplicationGatewaySslCertificateBase : NetworkBaseCmdlet
...
+               Mandatory = false,
+               HelpMessage = "SecretId (uri) of the KeyVault Secret. Use this option when a specific version of secret needs to be used.")]
+        [ValidateNotNullOrEmpty]
+        public string KeyVaultSecretId { get; set; }
+
+        [Parameter(
+               Mandatory = false,
+               HelpMessage = "KeyVault Secret Object. This option will use a version-less url for the certificate.")]
+        [ValidateNotNullOrEmpty]
+        public PSKeyVaultSecret KeyVaultSecret { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/SslCertificate/AzureApplicationGatewaySslCertificateBase.cs]: @akshaysngupta we can directly reference a type from another module -- this will cause a lot of issues and a coupling between the KeyVault and Network modules. The best thing we can do is have this `PSKeyVaultSecret` class inherit from a base class in common code that both KeyVault and Network have access to, but this will require a good amount of additional work. I think our best option is to avoid adding the `-KeyVaultSecret` parameter and just add the `-KeyVaultSecretId` parameter, which still exposes the functionality to the users.

File: src/ResourceManager/Network/Commands.Network/Commands.Network.Netcore.csproj [hunk 8]
```diff
@@ -68,4 +68,8 @@
   <ItemGroup>
     <Content Include="help\**\*" CopyToOutputDirectory="PreserveNewest" />
   </ItemGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\KeyVault\Commands.KeyVault\Commands.KeyVault.Netcore.csproj" />
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/Commands.Network.Netcore.csproj]: @akshaysngupta please remove this reference -- as mentioned before, we cannot introduce a coupling between modules

File: src/ResourceManager/Network/Commands.Network/Commands.Network.csproj [hunk 9]
```diff
@@ -58,9 +58,19 @@
       <HintPath>..\..\..\packages\Microsoft.Azure.KeyVault.Core.1.0.0\lib\net40\Microsoft.Azure.KeyVault.Core.dll</HintPath>
       <Private>True</Private>
     </Reference>
-    <Reference Include="Microsoft.Azure.Management.Network, Version=19.5.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.19.7.0-preview\lib\net452\Microsoft.Azure.Management.Network.dll</HintPath>
+    <Reference Include="Microsoft.Azure.Management.Network, Version=19.6.2.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.19.6.2-preview\lib\net452\Microsoft.Azure.Management.Network.dll</HintPath>
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/Commands.Network.csproj]: @akshaysngupta same comment about downgrading the version of the Network SDK

File: src/ResourceManager/Network/Commands.Network/Commands.Network.csproj [hunk 10]
```diff
@@ -58,9 +58,19 @@
...
     </Reference>
+    <Reference Include="Microsoft.Rest.ClientRuntime, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/Commands.Network.csproj]: @akshaysngupta same comment about removing this block of references as they are introduced elsewhere in common files

File: src/ResourceManager/Network/Commands.Network/Commands.Network.csproj [hunk 11]
```diff
@@ -801,6 +814,10 @@
     </EmbeddedResource>
   </ItemGroup>
   <ItemGroup>
+    <ProjectReference Include="..\..\KeyVault\Commands.KeyVault\Commands.KeyVault.csproj">
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/Commands.Network.csproj]: @akshaysngupta same comment about removing this reference

File: src/ResourceManager/Network/Commands.Network/packages.config [hunk 12]
```diff
@@ -1,6 +1,9 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
   <package id="AutoMapper" version="6.2.2" targetFramework="net452" />
   <package id="Microsoft.Azure.KeyVault.Core" version="1.0.0" targetFramework="net45" />
-  <package id="Microsoft.Azure.Management.Network" version="19.7.0-preview" targetFramework="net452" />
+  <package id="Microsoft.Azure.Management.Network" version="19.6.2-preview" targetFramework="net452" />
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/packages.config]: @akshaysngupta same comment about downgrading the Network SDK version

File: src/ResourceManager/Network/Commands.Network/packages.config [hunk 13]
```diff
@@ -1,6 +1,9 @@
...
+  <package id="Microsoft.Rest.ClientRuntime" version="2.3.18" targetFramework="net452" />
```
Comment (cormacpayne) [inline src/ResourceManager/Network/Commands.Network/packages.config]: @akshaysngupta same comment about removing the below three references

**Comment by number213** [comment]
Shouldn't this target `network-december-release`?

File: src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj (hunk 1 above)
Comment (akshaysngupta) [reply src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj]: removed.

File: src/ResourceManager/Network/Commands.Network.Test/packages.config (hunk 3 above)
Comment (akshaysngupta) [reply src/ResourceManager/Network/Commands.Network.Test/packages.config]: removed.

File: src/ResourceManager/Network/Commands.Network.Test/packages.config (hunk 5 above)
Comment (akshaysngupta) [reply src/ResourceManager/Network/Commands.Network.Test/packages.config]: removed.

File: src/ResourceManager/Network/Commands.Network.Test/packages.config (hunk 4 above)
Comment (akshaysngupta) [reply src/ResourceManager/Network/Commands.Network.Test/packages.config]: removed.

File: src/ResourceManager/Network/Commands.Network/Commands.Network.csproj (hunk 11 above)
Comment (akshaysngupta) [reply src/ResourceManager/Network/Commands.Network/Commands.Network.csproj]: removed.

File: src/ResourceManager/Network/Commands.Network/packages.config (hunk 13 above)
Comment (akshaysngupta) [reply src/ResourceManager/Network/Commands.Network/packages.config]: removed.

File: src/ResourceManager/Network/Commands.Network/packages.config (hunk 12 above)
Comment (akshaysngupta) [reply src/ResourceManager/Network/Commands.Network/packages.config]: removed.

File: src/ResourceManager/Network/Commands.Network/Commands.Network.csproj (hunk 10 above)
Comment (akshaysngupta) [reply src/ResourceManager/Network/Commands.Network/Commands.Network.csproj]: removed.

File: src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj (hunk 2 above)
Comment (akshaysngupta) [reply src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj]: removed.

**Comment by cormacpayne** [comment]
@akshaysngupta please resolve the merge conflicts

File: src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/Common.ps1 [hunk 14]
```diff
@@ -93,7 +93,7 @@ Cleans the created resource groups
 function Clean-ResourceGroup($rgname)
 {
     if ((Get-NetworkTestMode) -ne 'Playback') {
-        Remove-AzureRmResourceGroup -Name $rgname -Force
+        Remove-AzResourceGroup -Name $rgname -Force
```
Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/Common.ps1]: You don't want to do this - the resourcemanager script uses the old cmdlet names and aliases are turned on in all the tests

File: src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/ApplicationGatewayTests.ps1 [hunk 15]
```diff
@@ -97,102 +97,102 @@ function Test-ApplicationGatewayCRUD
 	try 
 	{
 		# Create the resource group
-		$resourceGroup = New-AzureRmResourceGroup -Name $rgname -Location $location -Tags @{ testtag = "APPGw tag"} 
+		$resourceGroup = New-AzResourceGroup -Name $rgname -Location $location -Tags @{ testtag = "APPGw tag"}
```
Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/ApplicationGatewayTests.ps1]: You should not make changes for existing cmdlets - only new cmdlets need to use the Az prefix

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs [hunk 16]
```diff
@@ -183,6 +183,13 @@ public class NewAzureApplicationGatewayCommand : ApplicationGatewayBaseCmdlet
             HelpMessage = "A hashtable which represents resource tags.")]
         public Hashtable Tag { get; set; }
 
+        [Parameter(
+            Mandatory = false,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "ResourceId of the user assigned identity to be assigned to Application Gateway.")]
+        [ValidateNotNullOrEmpty]
+        public string UserAssignedIdentityId { get; set; }
```
Comment (markcowl) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/NewAzureApplicationGatewayCommand.cs]: For parallelism with other cmdlets, this shoudl be UserAssignedIdentity - you could also add a parameter alias to allow both 'UserAssignedIdentity' and 'UserAssignedIdentityId'

**Comment by markcowl** [comment]
Closign in favor of #8115
